### PR TITLE
✨ Register API action for copy-to-clipboard function

### LINKED ISSUE #35467: I2I: Create an Bento component for copy to clipboard functionality
### Summary

This I2I proposes the `copy to clipboard` button. This button will allow users to copy any (specified in button attribute) content present in the document/page to the clipboard. 

### Design Document

We would like to introduce a new component `amp-copy` which will copy specified content into the clipboard.

### Attributes:

- `sourceId` - In this attribute, we can specify from where the content should get copied. This can be the `id` of any `HTML` tag (e.g `div`, `p`, `pre`, `code` etc.) from where we want to fetch the content. We can also provide the `id` of any `input` tag here.
- `text` - Any text passed to this attribute will get copied to the `clipboard`. Preference will be given to the `sourceId` attribute. If `sourceId` is not specified, we will look for the `text` attribute.

### Examples of both the attributes:

- Copy content from the `div` tag
 
```
<div id="myDiv1">
    <p> hello world </p>
</div>
<amp-copy width="100" height="20" source-id="myDiv1">
   Copy Content
</amp-copy>
```

- Copy content from the `text` tag

```
<amp-copy width="100" height="20" text="copy this line">
    Copy Content
</amp-copy>
```

This component doesn't expose a postMessage interface.

### Motivation

Currently, there's no direct way to add the `copy to clipboard` button on pages. We need a simple way to add such buttons on pages that will allow a wide range of users/publishers to allow their audience/users to copy content which just a single click.

### Alternative Solutions

There's a way to add the [`copy button`](https://amp.dev/documentation/examples/interactivity-dynamic-content/copy_button/) button on AMP pages. This solution involves adding an `iframe` to the page. Using `iframe` for copy seems to be overkill.

### Launch Tracker

_No response_

### Notifications

/cc @ampproject/wg-approvers

### PR DESCRIPTION
**Pull request by mangesh**
Fixes #35467

The goal of the PR is to support `copying to the clipboard` AMP action.

Introducing an action to facilitating clipboard copying which is useful not only for AMP.dev but also for any site that has code samples.

### Features
- Copy any content to the clipboard
   - Support for copying value of `input` tags (including `textarea`)
   - Support for copying text content inside an HTML tag (e.g `div`, `code`, `pre`, `p` tags)
- Custom events on copy action
   - `copy-success` event when content is successfully copied
   - `copy-error` event is fired when there's an error while copying the content
   - `support-error` event is fired when copying feature is not supported by browser

This action can be used on any HTML element

Examples:
```html
// Copy content from DIV
<div id='divTarget'>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</div>
<button on="tap:divTarget.copy();copy-success:divTarget.hide()">Copy DIV Content</button>
```

```html
// Copy text from input
<input id='inputTarget' value="Sample Input Value" />
<button on="tap:inputTarget.copy();copy-success:inputTarget.hide()">Copy Input Value</button>
```

```html
// Copy Static Text
<button on="tap:AMP.copy(text='Hello AMP');copy-success:copied.show()">Copy Static Text</button>
<p hidden id="copied">Copied!</p>
```

### In Progress
- Test cases

### Output/Demo
copy-to-clipboard

### CODE REVIEW

File: src/service/standard-actions-impl.js
Comment (rcebulko) [reply src/service/standard-actions-impl.js:204]: Yes 👍🏻

**Comment by mangesh** [comment]
@rcebulko Thanks for the feedback. I have fixed it. 👍

File: src/service/standard-actions-impl.js
Comment (caroqliu) [inline src/service/standard-actions-impl.js:227]: Can these two be collapsed? If the browser doesn't support the copy action, we can fire `copy-error` with `event.data.type = 'browser'` to indicate a support error.

Comment (mangesh) [reply src/service/standard-actions-impl.js]: This has been fixed.

**Review by caroqliu — COMMENTED** [review]
Thanks for the contribution! A few notes:
- Please pull in `main` to get CI to run: 
https://app.circleci.com/pipelines/github/ampproject/amphtml/15968/workflows/85176ba5-15fc-42ce-8335-8692f21010b2/jobs/281848
- Please update documentation in `docs/spec/amp-actions-and-events.md`
- Need feedback from @ampproject/wg-amp4email if this action would be allowable in the format -- if not, this should be disallowed by default already here: https://github.com/ampproject/amphtml/blob/ed3aaa73e12f022b10c6729a6c3f3bc1de7cdabf/src/service/action-impl.js#L47-L55

**Comment by zhangsu** [comment]
+1 for keeping it banned for AMP for Email. For security reasons we shouldn't allow the email sender to access any local storage of any sorts such as the system clipboard.

**Comment by dethstrobe** [comment]
Do you think we could get an automated unit or e2e test to test this functionality?

Other then that, it looks relatively straight forward. I think it does need to pull down main and resolve some merge conflicts, but it should be good to go after that.

**Review by dethstrobe — APPROVED** [review]

**Comment by AnuragVasanwala** [comment]
Thank you so much @dethstrobe.

> Do you think we could get an automated unit or e2e test to test this functionality?

Yes, I think it is good to have an automated unit test or e2e test. Let me have a look.

> Other then that, it looks relatively straight forward. I think it does need to pull down main and resolve some merge conflicts, but it should be good to go after that.

I have resolved conflicts and pulled latest from `main`.

Also, please let me know if it is required to update `OWNERS` for the PR: `ampproject/owners-check`

File: src/service/standard-actions-impl.js
Comment (jridgewell) [inline src/service/standard-actions-impl.js:262]: Nit: Move the `let` to here.

File: src/service/standard-actions-impl.js
Comment (jridgewell) [inline src/service/standard-actions-impl.js:273]: Is event value necessary?

**Review by jridgewell — CHANGES_REQUESTED** [review]
Need tests in `test/unit/test-action.js` before we can merge this.

File: src/service/standard-actions-impl.js
Comment (AnuragVasanwala) [reply src/service/standard-actions-impl.js:272]: @jridgewell 

It is necessary but not mandatory. Usually, `eventValue` is useful when when there is an error while copying. It provide three distinct value:
1. `success`: when copying action executed successfully
2. `error`: when there is an error while copying (i.e. end-user disabled copying for web-page, etc.)
3. `browser`: when browser do not support copy functionality

Please let me know your opinion whether to remove this functionality.

**Comment by AnuragVasanwala** [comment]
Need help for unit-test `test/unit/test-standard-actions.js`:
```javascript
describes.realWin('copy action', {amp: true}, (env) => {
  let ampdoc, standardActions, win;
  beforeEach(() => {
    ampdoc = new AmpDocSingle(window);
    env.sandbox.stub(AmpDocService.prototype, 'getAmpDoc').returns(ampdoc);
    standardActions = new StandardActions(ampdoc);
    win = env.win;
  });

  function trustedInvocation(obj) {
    return {satisfiesTrust: () => true, ...obj};
  }

  it('should copy from DIV', async () => {
    const divElement = win.document.createElement('div');
    divElement.textContent = 'Hello World!';

    const invocation = trustedInvocation({
      node: divElement,
      caller: divElement,
    });
    standardActions.handleCopy_(invocation);

    let text = '';
    console./* Review */ log(
      'unit-test iFrame hasFocus? : ',
      win.document.hasFocus()           // <-- Always returns `false`
    );

    win.navigator.clipboard
      .readText()
      .then((res) => {
        text = res;
      })
      .catch((e) => {
        /**
         * win.navigator.clipboard.readText() => always fails in unit-test with 'NotAllowedError: Document is not focused.' error.
         *
         * This is happening because unit-test is running inside an iFrame which is not focused.
         * Any workaround?
         */
        console./* Review */ error(e);
      });
    expect(text).to.equal('Hello World');
  });
});
```

I am trying to get copied value from clipboard but unit-test is performed inside an `iframe`, document of the `iframe` is not focused and `win.navigator.clipboard.readText()` always ends with `NotAllowedError: Document is not focused.` error.

Please provide necessary suggestions/corrections.

**Comment by dethstrobe** [comment]
@AnuragVasanwala Sorry it took so long to get back to you. I think what we want to do is spy on the document methods and make sure they're called correctly. While, I realize this isn't ideal as it doesn't prove functionality, however I do think it should help us avoid the issues with needing to have focus on the element as well as avoiding the browser blocking our action based on no user input.

I think we also will want to use the `window.navigator.clipboard` api and fallback to the `document.execCommand` if the clipboard api isn't available.

This should be a pretty decent unit test to cover this functionality.

```
describes.realWin('copy action', {amp: true}, (env) => {
  let ampdoc, standardActions, win;
  beforeEach(() => {
    ampdoc = new AmpDocSingle(window);
    env.sandbox.stub(AmpDocService.prototype, 'getAmpDoc').returns(ampdoc);
    standardActions = new StandardActions(ampdoc);
    win = env.win;
  });

  function trustedInvocation(obj) {
    return {satisfiesTrust: () => true, ...obj};
  }

  it('should copy from DIV using navigator.clipboard api', async () => {
    env.sandbox.spy(env.win.navigator.clipboard, 'writeText');
    const divElement = win.document.createElement('div');
    divElement.textContent = 'Hello World!';

    const invocation = trustedInvocation({
      node: divElement,
      caller: divElement,
    });
    standardActions.handleCopy_(invocation);

    await expect(env.win.navigator.clipboard.writeText).to.be.calledWith(
      'Hello World!'
    );
  });

  it('should fall back to legacy "doc.execCommand" if clipboard is not on the window.navigator', async () => {
    Object.defineProperties(env.win.navigator, {
      clipboard: {
        value: undefined,
        writable: true,
      },
    });

    env.sandbox.spy(env.win.document, 'execCommand');
    const divElement = win.document.createElement('div');
    divElement.textContent = 'Live long and prosper!';

    const invocation = trustedInvocation({
      node: divElement,
      caller: divElement,
    });
    standardActions.handleCopy_(invocation);

    await expect(env.win.document.execCommand).to.be.calledWith('copy');
  });
});
```

**Review by dethstrobe — APPROVED** [review]
Is there no test?

**Comment by AnuragVasanwala** [comment]
> Is there no test?

@dethstrobe Yes, there is unit-test for `copy` action. I am working on it as per [your suggestion](https://github.com/ampproject/amphtml/pull/35799#issuecomment-1021466288). It will be ready soon 🙂

**Comment by AnuragVasanwala** [comment]
> Need tests in `test/unit/test-action.js` before we can merge this.

@jridgewell & @dethstrobe I have worked on feedback and made necessary changes:
- Support for `navigator.clipboard` API
- Support for `document.execCommand('copy')` deprecated API as a fallback for older browser without support for `navigator.clipboard` API
- Unit-test for each `copy` action (`test/unit/test-standard-actions.js`)
  - Copy Static Text
  - Copy from DIV Content
  - Copy from INPUT Value

Please review and let me know in case of any issue.

File: src/service/standard-actions-impl.js
Comment (dethstrobe) [inline src/service/standard-actions-impl.js:222]: We don't really need this to have a return type, right?

File: src/core/window/clipboard.js
Comment (rcebulko) [inline src/core/window/clipboard.js:11]: Given this is a non-async op, why the change to use callbacks? Also, if we stick with callbacks, we should definitely type them accordingly rather than using `*` any

File: src/service/standard-actions-impl.js
Comment (rcebulko) [reply src/service/standard-actions-impl.js:222]: Also the `!` prefix doesn't do anything/should be left out now that we've switched off of CC.

File: src/service/standard-actions-impl.js
Comment (jridgewell) [inline src/service/standard-actions-impl.js:289]: Nit: The reused variables aren't really necessary, and the code will minify without them. I would just call the trigger event with the correct values, and remove the variables.

File: src/service/standard-actions-impl.js
Comment (jridgewell) [inline src/service/standard-actions-impl.js:298]: Can you explain the browser event result? Should this be `unsupported` instead?

File: src/service/standard-actions-impl.js
Comment (jridgewell) [reply src/service/standard-actions-impl.js:222]: Yah, we can just remove this and the return statement.

File: src/core/window/clipboard.js
Comment (jridgewell) [inline src/core/window/clipboard.js:79]: Nit: 
```suggestion
  if (win.navigator?.clipboard) {
```

File: src/core/window/clipboard.js
Comment (jridgewell) [inline src/core/window/clipboard.js:91]: Can we just pass the callsbacks directly to `then`?
```suggestion
      successCallback(),
      failCallback()
```

**Review by jridgewell — APPROVED** [review]

File: src/core/window/clipboard.js
Comment (rcebulko) [reply src/core/window/clipboard.js:91]: no parens if passed directly
```suggestion
        successCallback, failCallback
```

File: src/core/window/clipboard.js
Comment (rcebulko) [reply src/core/window/clipboard.js:11]: It looks to me like this is fine returning a boolean and the calling code can call callbacks as needed?

File: src/service/standard-actions-impl.js
Comment (AnuragVasanwala) [reply src/service/standard-actions-impl.js:222]: Sure, I will correct it.

File: src/core/window/clipboard.js
Comment (AnuragVasanwala) [reply src/core/window/clipboard.js:11]: > It looks to me like this is fine returning a boolean and the calling code can call callbacks as needed based on that return value?

I think you are correct. This approach seems much clear. I will revert it back to the original function returning `boolean` and the calling code can call `callbacks` as needed based on the return value.

File: src/service/standard-actions-impl.js
Comment (AnuragVasanwala) [reply src/service/standard-actions-impl.js:288]: Absolutely! Your approach seems more clear. I will update accordingly.

File: src/service/standard-actions-impl.js
Comment (AnuragVasanwala) [reply src/service/standard-actions-impl.js:297]: Yes, sorry the confusing name. I have renamed it to `unsupported` instead.

File: src/core/window/clipboard.js
Comment (AnuragVasanwala) [reply src/core/window/clipboard.js]: Thank you for highlighting. I have corrected these calls to callback.

File: src/core/window/clipboard.js
Comment (rcebulko) [inline src/core/window/clipboard.js:90]: A few comments to simplify this a bit:
1. Comments of the form `/** ... */` typically indicate JSDoc format doc comments, whereas `//`-style comments are more typically used in this repo for inline explanatory comments
2. Combining the checks here could be intuitive and reduces a layer of `if` nesting
3. Since the anonymous functions being passed to the `.then(...)` here do nothing but call the success/failure callbacks, those callbacks can be passed directly.

```javascript
export function copyTextToClipboard(win, text, successCallback, failCallback) {
  if (!!win.navigator?.clipboard) {
    win.navigator.clipboard.writeText(text).then(successCallback, failCallback);
  } else if (isDeprecatedCopyingToClipboardSupported(win.document) &&
             deprecatedCopyTextToClipboard(win, text)) {
    successCallback();  
  } else {
    failCallback();
  }
}
```

File: src/service/standard-actions-impl.js
Comment (rcebulko) [inline src/service/standard-actions-impl.js:292]: See above re: JSDoc-style comments. Since the first ~line of this is describing the following line of code, it may not be adding any new information.
```suggestion
          // Error encountered while copying.
```

File: src/service/standard-actions-impl.js
Comment (rcebulko) [inline src/service/standard-actions-impl.js:300]:
```suggestion
      // Copy is disabled or not supported by user.
```

File: src/core/window/clipboard.js
Comment (AnuragVasanwala) [reply src/core/window/clipboard.js:78]: > A few comments to simplify this a bit:
> 
> 1. Comments of the form `/** ... */` typically indicate JSDoc format doc comments, whereas `//`-style comments are more typically used in this repo for inline explanatory comments
> 2. Combining the checks here could be intuitive and reduces a layer of `if` nesting
> 3. Since the anonymous functions being passed to the `.then(...)` here do nothing but call the success/failure callbacks, those callbacks can be passed directly.
> 
> ```js
> export function copyTextToClipboard(win, text, successCallback, failCallback) {
>   if (!!win.navigator?.clipboard) {
>     win.navigator.clipboard.writeText(text).then(successCallback, failCallback);
>   } else if (isDeprecatedCopyingToClipboardSupported(win.document) &&
>              deprecatedCopyTextToClipboard(win, text)) {
>     successCallback();  
>   } else {
>     failCallback();
>   }
> }
> ```

Resolved!

File: src/service/standard-actions-impl.js
Comment (rcebulko) [inline src/service/standard-actions-impl.js:238]:
```suggestion
      // Copy Static Text
      //  Example: AMP.copy(text='TextToCopy');
```

File: src/service/standard-actions-impl.js
Comment (rcebulko) [inline src/service/standard-actions-impl.js:244]:
```suggestion
      // Copy Target Element Text
      //  Example: targetId.copy();
```

**Review by rcebulko — APPROVED** [review]

**Comment by dethstrobe** [comment]
@AnuragVasanwala Thanks for the PR and all the hard work. I'm going to merge this PR as I feel holding up this PR any longer over code styling comments is unnecessary. And we can always refactor it later if we want.

Thank you again for all the hard work.

**Pull request merged**
